bootstrap/eks/internal/userdata/nodeadm.go

```diff
@@ -0,0 +1,327 @@
+	}
+	if _, ok := labelsMap[nodeLabelCapacityType]; !ok {
+		labelsMap[nodeLabelCapacityType] = input.getCapacityTypeString()
+	}
+	stringBuilder := strings.Builder{}
+	for key, value := range labelsMap {
+		stringBuilder.WriteString(fmt.Sprintf("%s=%s,", key, value))
+	}
+	newLabels := stringBuilder.String()[:len(stringBuilder.String())-1] // remove the last comma
+	newFlags = append(newFlags, fmt.Sprintf("--node-labels=%s", newLabels))
+	input.KubeletFlags = newFlags
+	return nil
+}
+
+// getCapacityTypeString returns the string representation of the capacity type.
+func (input *NodeadmInput) getCapacityTypeString() string {
+	if input.CapacityType == nil {
+		return "ON_DEMAND"
+	}
+	switch *input.CapacityType {
+	case v1beta2.ManagedMachinePoolCapacityTypeSpot:
+		return "SPOT"
+	case v1beta2.ManagedMachinePoolCapacityTypeOnDemand:
+		return "ON_DEMAND"
+	default:
+		return strings.ToUpper(string(*input.CapacityType))
+	}
+}
+
+// validateNodeInput validates the input for nodeadm user data generation.
+func validateNodeadmInput(input *NodeadmInput) error {
+	if input.APIServerEndpoint == "" {
+		return fmt.Errorf("API server endpoint is required for nodeadm")
+	}
+	if input.CACert == "" {
+		return fmt.Errorf("CA certificate is required for nodeadm")
+	}
+	if input.ClusterName == "" {
+		return fmt.Errorf("cluster name is required for nodeadm")
+	}
+	if input.NodeGroupName == "" {
+		return fmt.Errorf("node group name is required for nodeadm")
+	}
+	if input.Boundary == "" {
+		input.Boundary = boundary
+	}
+	err := input.setKubeletFlags()
+	if err != nil {
+		return err
+	}
+
+	klog.V(2).Infof("Nodeadm Userdata Generation - node-labels: %s", input.NodeLabels)
+
+	return nil
+}
+
+// NewNodeadmUserdata returns the user data string to be used on a node instance.
+func NewNodeadmUserdata(input *NodeadmInput) ([]byte, error) {
+	if err := validateNodeadmInput(input); err != nil {
+		return nil, err
+	}
+
+	var buf bytes.Buffer
+
+	// Write MIME header
+	if _, err := buf.WriteString(fmt.Sprintf("MIME-Version: 1.0\nContent-Type: multipart/mixed; boundary=%q\n\n", input.Boundary)); err != nil {
+		return nil, fmt.Errorf("failed to write MIME header: %v", err)
+	}
+
+	// Write shell script part if needed
+	if len(input.PreBootstrapCommands) > 0 {
+		shellScriptTemplate := template.Must(template.New("shell").Parse(shellScriptPartTemplate))
+		if err := shellScriptTemplate.Execute(&buf, input); err != nil {
+			return nil, fmt.Errorf("failed to execute shell script template: %v", err)
+		}
+		if _, err := buf.WriteString("\n"); err != nil {
+			return nil, fmt.Errorf("failed to write newline: %v", err)
+		}
+	}
+
+	// Write node config part
+	nodeConfigTemplate := template.Must(
+		template.New("node").
+			Funcs(defaultTemplateFuncMap).
+			Parse(nodeConfigPartTemplate),
+	)
+	if err := nodeConfigTemplate.Execute(&buf, input); err != nil {
+		return nil, fmt.Errorf("failed to execute node config template: %v", err)
+	}
+
+	// Write cloud-config part
```

Comment:
The cloud-init executes all three parts in sequence. The shell-script part is added first then `nodeconfig` and then `cloud-init`. The files are added in the cloud-config part. How the shell script part will execute the `PreBootstrapCommands` if the files that it refers are not dropped in yet? 
Can you please explain how it works?

Also if the `shell script` part reboots, does cloud-init knows not to run it again after reboot and continue with next part?

Reply:
i dont know _how_ it works. those are cloud-init internals, but based on logs from cloud-init it works.

Reply:
ok. we can test it out and change ordering in followup PR if needed.